pipeline/routing/ppr.py

```diff
@@ -33,56 +33,93 @@ def insertPathsElementsRef(cur, edges, path_counter):
+    
+    # Logical structure of the id generation for access spaces:
+    # level:                       0                   1                     0                   -1                     0
+    # OSM:            (id_1) ---Footway--- (id_2) ---Stairs--- (id_3) ---Elevator--- (id_3) ---Footway--- (id_4) ---Escalator--- (id_5)
+    # access spaces:  ----------- [X] --------------- [X] ----------------- [X] ---------------- [X] ----------------- [X] ------------
+    # id(id,level):   ----------------- (id_2,NULL) --------- (id_3,1) ----------- (id_3,-1) --------- (id_4,NULL) --------------------
+    
+    # Special cases:
+    
+    # Elevators:
+    # Their 'osm_way_id', 'from_node_osm_id' and 'to_node_osm_id' are the same. One elevator can have multiple levels, and therefore multiple access spaces.
+    # So the elevators access spaces are identified by the level of the previous edge when stepping into the elevator,
+    # and the level of the current edge when stepping out of the elevator.
+    
+    # Escalators:
+    # Their level is dependent on the direction of the path. So the access spaces are identified by the level of the previous edge when going into the escalator,
+    # and the level of the current edge when going out of the escalator.
+    
+    # Stairs:
+    # They always have the same level, regardless of the direction. So the access spaces are identified by the level of the previous edge when going into the stairs,
+    # and the level of the current edge when going out of the stairs.
     
     for edge in edgeIter:
         osm_way_id = abs(edge["osm_way_id"])
         edge_type = edge["edge_type"]
         street_type = edge["street_type"]
         from_node_osm_id = edge["from_node_osm_id"]
         to_node_osm_id = edge["to_node_osm_id"]
+        level = edge["level"]
 
         if( # 1) transition from one edge_type to another
             edge_type != previous_edge_type and
             (edge_type in special_edge_types or previous_edge_type in special_edge_types)
-        )\
-        or( # 2) transition from one street_type to another
-            street_type != previous_street_type and
-            (street_type in special_street_types or previous_street_type in special_street_types)
-        ):
-            insertAccessSpaces(cur, from_node_osm_id, 'N', None, None, edge["path"][0])
-            previous_edge_type = edge_type
-            previous_street_type = street_type
+            )\
+            or( # 2) transition from one street_type to another
+                street_type != previous_street_type and
+                (street_type in special_street_types or previous_street_type in special_street_types)
+            ):
+            # special case: elevator
+            if edge_type == "elevator" or street_type == "stairs" or street_type == "escalator":
+                # going into the elevator/stairs: use level from the previous edge
+                current_level = previous_level
+            else:
```

Comment:
Maybe we can note here, that this approach might fail if two of these are directly connected (lets say stairs directly follow en escalator). At least this is what I remember.

Reply:
Comment added in [ba9dcbd](https://github.com/OPENER-next/osm2vdv462/commit/ba9dcbdd96b85776bc3d53e8c6226ba333b0de73)